File: frontend/src/microservices/operations/train/api/RegisterRailwayManagerPlanner.ts
import { executeJsonApiRequest } from '@/shared-kernel/api/ApiTransport'
import type { TrainAdminSessionPlannerResponse } from '@/microservices/operations/train/objects/TrainAdminSessionPlannerResponse'
import type { RegisterRailwayManagerPlannerRequest } from '@/microservices/operations/train/objects/RegisterRailwayManagerPlannerRequest'

export const registerRailwayManagerPlanner = (payload: RegisterRailwayManagerPlannerRequest): Promise<TrainAdminSessionPlannerResponse> =>
  executeJsonApiRequest('/RegisterRailwayManagerPlanner', 'POST', payload)

export const registerRailwayManager = registerRailwayManagerPlanner

File: backend/src/main/scala/microservices/operations/train/api/RegisterRailwayManagerPlanner.scala
package com.typesafe.travel.operations.domain

import cats.effect.IO
import com.typesafe.travel.api.routes.ConnectionApiPlan
import com.typesafe.travel.auth.domain.hashPasswordForLoginEmail
import com.typesafe.travel.shared.kernel.EmailAddress

import java.sql.Connection

object RegisterRailwayManagerPlanner extends ConnectionApiPlan[RegisterRailwayManagerPlannerRequest, TrainAdminSessionPlannerResponse]:
  override val name: String = "RegisterRailwayManagerPlanner"
  override def plan(input: RegisterRailwayManagerPlannerRequest, connection: Connection): IO[TrainAdminSessionPlannerResponse] =
    for
      email <- IO.fromEither(EmailAddress.create(input.email))
      passwordHash <- hashPasswordForLoginEmail(input.password, email)
      response <- TrainManagerPlainSql.registerManager(connection, input, passwordHash, java.time.Instant.now())
    yield response
